### PR TITLE
setoid_rewrite: record generated name when rewriting under lambda 

### LINKED ISSUE #13129: setoid_rewrite does not work with name mangling
#### Description of the problem

Under some circumstances, setoid_rewrite fails when name mangling is enabled. Here is a reproducible example:
```coq
From Coq Require Export Morphisms Setoid.

Set Mangle Names.

(** ** Setoid equality *)
(** We define an operational type class for setoid equality, i.e., the
"canonical" equivalence for a type. The typeclass is tied to the \equiv
symbol. This is based on (Spitters/van der Weegen, 2011). *)
Class Equiv A := equiv: relation A.
(* No Hint Mode set because of Coq bug #5735
Hint Mode Equiv ! : typeclass_instances. *)

Infix "≡" := equiv (at level 70, no associativity).
Infix "≡@{ A }" := (@equiv A _)
  (at level 70, only parsing, no associativity).

Notation "(≡)" := equiv (only parsing).

(** Unbundled version *)
Class Dist A := dist : nat -> relation A.
Instance: Params (@dist) 3 := {}.
Notation "x ≡{ n }≡ y" := (dist n x y)
  (at level 70, n at next level, format "x  ≡{ n }≡  y").
Notation "x ≡{ n }@{ A }≡ y" := (dist (A:=A) n x y)
  (at level 70, n at next level, only parsing).

Hint Extern 0 (_ ≡{_}≡ _) => reflexivity : core.
Hint Extern 0 (_ ≡{_}≡ _) => symmetry; assumption : core.
Notation NonExpansive f := (forall n, Proper (dist n ==> dist n) f).
Notation NonExpansive2 f := (forall n, Proper (dist n ==> dist n ==> dist n) f).

Record OfeMixin A `{Equiv A, Dist A} := {
  mixin_equiv_dist (x y : A) : x ≡ y <-> forall n, x ≡{n}≡ y;
  mixin_dist_equivalence n : Equivalence (@dist A _ n);
  mixin_dist_S n (x y : A) : x ≡{S n}≡ y -> x ≡{n}≡ y
}.

(** Bundled version *)
Structure ofeT := OfeT {
  ofe_car :> Type;
  ofe_equiv : Equiv ofe_car;
  ofe_dist : Dist ofe_car;
  ofe_mixin : OfeMixin ofe_car
}.
Arguments OfeT _ {_ _} _.
Add Printing Constructor ofeT.
Hint Extern 0 (Equiv _) => eapply (@ofe_equiv _) : typeclass_instances.
Hint Extern 0 (Dist _) => eapply (@ofe_dist _) : typeclass_instances.
Arguments ofe_car : simpl never.
Arguments ofe_equiv : simpl never.
Arguments ofe_dist : simpl never.
Arguments ofe_mixin : simpl never.

(** Lifting properties from the mixin *)
Section ofe_mixin.
  Context {A : ofeT}.
  Implicit Types x y : A.
  Lemma equiv_dist x y : x ≡ y <-> forall n, x ≡{n}≡ y.
  Proof. apply (mixin_equiv_dist _ (ofe_mixin A)). Qed.
End ofe_mixin.

(** General properties *)
Section ofe.
  Context {A : ofeT}.
  Implicit Types x y : A.

  Lemma ne_proper_2 {B C : ofeT} (f : A -> B -> C) `{!NonExpansive2 f} :
    Proper ((≡) ==> (≡) ==> (≡)) f.
  Proof.
     unfold Proper, respectful. setoid_rewrite equiv_dist.
```
The last command works when the `Set Mangle Names` in the beginning is removed.

#### Coq Version

214215e5

### PR DESCRIPTION
**Pull request by SkySkimmer**
Fix #13129

### CODE REVIEW

File: plugins/ltac/rewrite.ml [hunk 1]
```diff
@@ -1131,31 +1117,13 @@ let subterm all flags (s : 'a pure_strategy) : 'a pure_strategy =
    B. Barras' idea is to have a context of relations, of length 1, with Σ for gluing
    dependent relations and using projections to get them out.
  *)
```
Comment (SkySkimmer) [inline plugins/ltac/rewrite.ml:1119]: BTW is this comment still TODO or is it done? cc @mattam82

**Comment by SkySkimmer** [comment]
Not sure how to make a smaller test even knowing where the bad code was.

**Comment by RalfJung** [comment]
Any update on this PR, anyone able to review it? Would be nice to move forward with name mangling in Iris, which is blocked on this fix. :)

**Review by ppedrot — APPROVED** [review]
The fix looks correct to me indeed.

**Comment by ppedrot** [comment]
Merging soon.

**Pull request merged**